auth/src/main/java/com/firebase/ui/auth/ui/provider/GitHubLoginActivity.java

```diff
@@ -0,0 +1,92 @@
+package com.firebase.ui.auth.ui.provider;
+
+import android.annotation.SuppressLint;
+import android.content.Context;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+import android.support.customtabs.CustomTabsIntent;
+import android.support.v4.content.ContextCompat;
+
+import com.firebase.ui.auth.R;
+import com.firebase.ui.auth.data.remote.GitHubSignInHandler;
+import com.firebase.ui.auth.ui.HelperActivityBase;
+import com.firebase.ui.auth.util.ExtraConstants;
+
+@SuppressLint("GoogleAppIndexingApiWarning")
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class GitHubLoginActivity extends HelperActivityBase {
+    private static final String REFRESH_ACTION = "refresh_action";
+
+    private boolean mShouldCloseCustomTab;
+
+    @NonNull
+    public static Intent createIntent(Context context, Uri starter) {
+        return new Intent(context, GitHubLoginActivity.class)
+                .putExtra(ExtraConstants.PARAMS, starter);
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (savedInstanceState == null) {
+            new CustomTabsIntent.Builder()
+                    .setShowTitle(true)
+                    .enableUrlBarHiding()
+                    .setToolbarColor(ContextCompat.getColor(this, R.color.colorPrimary))
+                    .build()
+                    .launchUrl(this, (Uri) getIntent().getParcelableExtra(ExtraConstants.PARAMS));
+            mShouldCloseCustomTab = false;
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        if (mShouldCloseCustomTab) { // User pressed back
+            finish(RESULT_CANCELED, null);
+        }
+        mShouldCloseCustomTab = true;
```

Comment:
Can you explain a little why this is set to `true` on each `onResume`?

Reply:
Yeah, I'll explain this and the recursive launch comment below. Basically, I've copied-ish all of this from Facebook since they actually did a really nice job with CCT. So here's what we're trying to do:
- Launch CCT
- If user presses back, close resources related to CCT
- Same for success and failure, but send result data too

Given that CCT is going to redirect to our activity, we need a wrapper with special stuff like `singleTop` and ignored config changes so the stack doesn't nest itself. Now that we're guaranteed to have a single activity with a CCT layer on top, we can safely assume that `onCreate` is the only place to start CCT. So the current flow looks like this:
- Launch CCT in `onCreate`
- Receive redirects in `onNewIntent`

That creates a problem though: how do we close CCT? Android doesn't give you a nice way to close all activities on top of the stack AFAIK, so we're force to relaunch our wrapper activity with the `CLEAR_TOP` flag. That will recurse while killing CCT to bring us back to `onNewIntent` again where we check for the refresh action. At that point, we can finally finish with our result.

Now for the `onResume` stuff. Remember how we always have a CCT layer on top? That means `onResume` will only ever be called once... unless the user presses back. At that point, our wrapper activity gains focus for the second time and we can safely kill it knowing it was a back event.

Whew! 😌 I realize that's a painful explanation. I dunno, Facebook's code was even more complicated and took me forever to even distill down to this point. They had 2 activities they were juggling around—never understood why. Anyway, if you have any ideas to simplify this stuff, please send them over! 😄

Reply:
That's all super cool!  Can you just encode some of this in code comments?  If you want you can literally copy what you just said to a class-level comment.